Sources/SwiftProtobuf/Message.swift

```diff
@@ -110,6 +110,9 @@ public protocol Message: _CommonMessageConformances {
   /// normal `Equatable`. `Equatable` is provided with specific generated
   /// types.
   func isEqualTo(message: Message) -> Bool
+    
+  /// Provides `Field` information for this `Message` type used to provide a default implementation of `traverse`
+  static var _fields: [Field<Self>] { get }
```

Comment:
In working with the Swift standard library team, we've learned that generic types are not free at runtime -- in particular, the runtime typically ends up having to actually create a runtime representation for each generic type.  So `Field<M>` will end up producing a separate runtime type representation for every message type `M` (much the same as if you had created a type `Field_MessageName` instead).  So it can make sense to combine or otherwise minimize the number of different generic types in a library where memory usage might be an issue.

I'm not sure there's a good alternative in this case: `Field<>` does seem like it needs to be generic over the message type.

Reply:
Agreed, I think the reified generics have a negative impact on binary size. IDK what kind of impact they have on memory but its probably a negative one too.

I did have a different approach I explored where I created a non-generic and less safe version of Field:
```swift
public struct UnsafeField {
    private let item: UnsafeFieldItem
    func unsafeTraverse<M: Message, V: Visitor>(message: M, visitor: inout V) throws {
        try item.unsafeTraverse(message: message, visitor: &visitor)
    }
    
    public static func singularFloat<M: Message>(_ message: M.Type, _ getValue: @escaping (M) -> Float, fieldNumber: Int, defaultValue: Float = 0) -> Self {
        Self(item: UnsafeSingularFloatField(fieldNumber: fieldNumber, getValue: getValue, isDefault: { getValue($0) == defaultValue }))
    }
}

private protocol UnsafeFieldItem {
    func unsafeTraverse<M: Message, V: Visitor>(message: M, visitor: inout V) throws
}

private protocol UnsafeGetterProtocol<ReturnValue> {
    associatedtype ReturnValue
    func invoke<M: Message>(message: M, then action: (ReturnValue) throws -> Void) throws
}

private struct UnsafeGetter<Msg: Message, ReturnValue>: UnsafeGetterProtocol {
    let getter: (Msg) -> ReturnValue
    let shouldSkip: (Msg) -> Bool

    func invoke<M: Message>(message: M, then action: (ReturnValue) throws -> Void) throws {
        let msg = unsafeBitCast(message, to: Msg.self)
        if !shouldSkip(msg) {
            return try action(getter(msg))
        }
    }
}

private struct UnsafeSingularFloatField: UnsafeFieldItem {
    private let fieldNumber: Int
    private let getter: any UnsafeGetterProtocol<Float>
    
    fileprivate init<M: Message>(fieldNumber: Int, getValue: @escaping (M) -> Float, isDefault: @escaping (M) -> Bool) {
        self.fieldNumber = fieldNumber
        self.getter = UnsafeGetter(getter: getValue, shouldSkip: isDefault)
    }
    
    func unsafeTraverse<M: Message, V: Visitor>(message: M, visitor: inout V) throws {
        try getter.invoke(message: message) {
            try visitor.visitSingularFloatField(value: $0, fieldNumber: fieldNumber)
        }
    }
}
```

This had identical runtime performance as the `Field<>` solution. However I think this ultimately still created the same amount of generic types and just pushed this cost to `UnsafeGetter` and  I didn't see noticeable difference in binary size impact.

The `UnsafeGetter` design was done this way to allow me to do `unsafeBitCast` which doesn't have a runtime penalty compared to doing an `as!`

However if the goal is to remove as many generics as possible `UnsafeGetter` could be done like this:
```swift
private struct UnsafeGetter<ReturnValue>: UnsafeGetterProtocol {
    
    let invokeIfShouldNotSkip: (any Message, (ReturnValue) throws -> Void) throws -> Void
    
    init<M: Message>(getter: (M) -> ReturnValue, shouldSkip: (M) -> Bool) {
        invokeIfShouldNotSkip = { msg, callback in
            let message = msg as! M
            if !shouldSkip(msg as! M) {
                try callback(getter(message))
            }
        }
    }

    func invoke<M: Message>(message: M, then action: (ReturnValue) throws -> Void) throws {
        try invokeIfShouldNotSkip(message, action)
    }
}
```

Reply:
> Does this actually work

Yes it works but only if the types have the exact same size in memory. It does not work if `message` is an `Any` which is why `UnsafeGetter` captured `Msg: Message` and the function captures `M: Message`. Otherwise the only way to do the cast is with an `as?`/`as!`

Reply:
Have you tested this for cases where the message has indirect storage?   In that case, the message has a reference-counted pointer to the class that actually stores the data, and I wasn't aware it was safe to use `unsafeBitCast` to copy a struct that has reference-counted pointers.

Reply:
I'm not sure, I'd have to dig up my old branch and see. Either way the approach that used `unsafeBitCast` is probably not the right way to go since in the end it created the same amount of generic classes due to `UnsafeGetter<Msg: Message, ReturnValue>` still capturing the message type. If the goal is reducing the amount of generics then the second approach I posted with `struct UnsafeGetter<ReturnValue>` would be better even thought the `as!` is less performant.

Reply:
I want to reduce memory usage without hurting performance "too much."   How much slower is your second approach?

Reply:
The difference is significant. Around 2x slower with `as!`